src/v/kafka/server/connection_context.h

```diff
@@ -77,6 +77,23 @@ class connection_context final
     ss::net::inet_address client_host() const { return _client_addr; }
 
 private:
+    // used to track number of pending requests
+    class request_tracker {
```

Comment:
Might be nice to have a helper:
```c++
auto track_request(rpc::server_probe& probe) {
    return std::make_unique<request_tracker>(probe);
}
```

It''s also possible to implement it by using a custom deleter directly in `std::unique_ptr`, something like:
```c++
auto track_request(rpc::server_probe& probe) {
    auto raii = [&probe]() noexcept {_probe.request_completed();}
    _probe.request_received();
    return std::make_unique<request_tracker, decltype(raii)>(probe, raii);
}
```
Which requires no allocation.

Reply:
In that second formulation, is it possible to express the full type elsewhere? e.g. what would the `tracker` member of connection_context look like?

Reply:
Good question. A no capture lambda can decay to a function pointer. Something like:
```
std::unique_ptr<rpc::server_probe, void(rpc::server_probe*)>
```
It may require some coercion instead of the `decltype`.